ouroboros-consensus/docs/utxo-hd/report.tex

```diff
@@ -0,0 +1,741 @@
+  ShelleyLedgerTables {
+    shelleyUTxOTable :: mk (SL.TxIn (EraCrypto era)) (Core.TxOut era)
+  }
+\end{lstlisting}
+
+\subsection{Cardano ledger tables}
+
+The Cardano ledger tables are the most problematic of all, as we chose not to
+define compositional tables.
+
+On the one hand, a ledger state of the Cardano block is a \texttt{Telescope} of
+ledger states of the eras so what we have in reality is only one of those era's
+ledger states. Therefore the most reasonable way of defining the tables would be
+to leave the tables to be defined by the NS-wrapped ledger state.
+
+On the other hand, a Cardano block is itself a block, so in order to have a
+uniform interface, it has to be able to use all the operations of blocks and in
+particular must have the concept of \emph{LedgerTables} associated with a block.
+On this regard, we defined the \texttt{CardanoledgerTables}:
+
+\begin{lstlisting}
+newtype LedgerTables (LedgerState (CardanoBlock c)) mk = CardanoLedgerTables {
+  cardanoUTxOTable :: mk (SL.TxIn c) (CardanoTxOut c)
+  }
+\end{lstlisting}
+where \texttt{CardanoTxOut} is an NS over the \texttt{TxOut} of the Shelley
+eras.
+
+There is a tension here that is resolved by projecting and injecting only at
+certain points in the execution.
+
+So whenever we are executing functions that deal with \emph{Cardano hard fork
+  blocks}, we will be carrying around \texttt{CardanoLedgerTables} and whenever
+specific instances call the concrete block instances, we will inject the values
+as tables of the concrete block type.
+
+In particular the DbChangelog for the Cardano block will carry differences which
+are \texttt{CardanoLedgerTables} and, on the other side of the spectrum, when we
+want to apply a concrete block to a concrete ledger state, we will inject the
+\texttt{CardanoLedgerTables} into the concrete block tables.
+
+\subsection{Era transitions}
+
+\subsubsection{Byron to Shelley}
+
+As Byron doesn't have tables but Shelley does, the era transition creates out of
+thin air a set of differences which insert all and every value on the UTxO set
+into the ledger tables. This huge set of differences will be eventually garbage
+collected once the first Shelley block ends up further than $k$ blocks from the
+node's tip and a flushing is performed.
+
+\subsubsection{Shelley to Allegra}
+
+On Byron, some addresses existed there to be reclaimed by
+investors\footnote{These were called AVVM addresses. Perhaps the Ledger team can
+  tell more about them}, and as some were never reclaimed, the ledger made the
+decision of giving the Ada in those addresses back to the treasury. This
+happened on the Allegra boundary, and therefore the ledger state has been
+modified to keep track of this information so that the era transition can
+consume these entries into the treasury and return a set of deletions of those
+addresses on the UTxO set.
+
+See \texttt{Cardano.Ledger.Allegra.Translation.shelleyToAllegraAVVMsToDelete}.
```

Comment:
In the first version of the anti-diff integration where forwarding would still perform sanity checks on the inserts/deletes, this translation proved to be a bit hairy. In particular, the translation would emit the AVVM deletions as diffs, but flushing these diffs to the in-memory backing store would error. My understanding was that it would error because the diffs declared values to be deleted even though these values were not present in the backing store. 

https://github.com/input-output-hk/ouroboros-network/tree/jdral/bugfix-shelley-allegra-translation fixes this issue by not emitting the deletions as diffs. I'm not sure if this is a correct fix though

Reply:
> My understanding was that it would error because the diffs declared values to be deleted even though these values were not present in the backing store.

This is wrong. The values *are* in the backing store